### DISCUSSION TITLE
Connection Specs for OpenCPN to UDP Receive

### DISCUSSION THREAD

**Discussion by AlanH70** [transmitterdan/VDRplayer#6] [Q&A] [answered]
Thanks to rgleason for aiming me at this project and to transmitterdan for providing all this work.  Nice!

The VDRplayer command line options seem clear.  To play an NMEA txt file created with the ShipDriver plugin [all !AIVDM sentences] to OpenCPN on the same windows machine I would expect to command line specify  UDP, IP addr = 127.0.0.1, and Port 2947, (and some other options).

But what do I specify in OpenCPN's Tools > Options > Connections for the UDP Address and Dataport to enable receiving what the python script is transmitting?

Alan

**Comment by rgleason** [comment]
Perhaps this will help, as an example.
https://opencpn.org/wiki/dokuwiki/doku.php?id=opencpn:manual_basic:set_options:connections:troubleshoot_with_putty

**Comment by transmitterdan** [comment]
UDP can use a multicast address or unicast address.  Assume you wish to use multicast and your network is 192.168.1.xxx (typical local home network).  This command:
```
./VDRplayer.py PacCupStartLong.txt --UDP --port=10110 --dest 192.168.1.255 --sleep=0.07 --repeat=100000
```
Would multicast to any device on the network.  On a computer connected to this network (say its IP address is 192.168.1.18 but any IP on this network will work) you would create this network connection:

![image](https://github.com/transmitterdan/VDRplayer/assets/10525699/45c5bb9a-297c-4608-983b-43398263bbb2)

Then OpenCPN will receive the UDP messages.

While UDP is "simple", I find TCP preferrable but it's a personal choice in most cases.

**Reply by AlanH70** [reply]
Just saw this.  Thanks Dan.  
My OpenCPN 5.8.4-0 connect window looks a bit different, with no Data Protocol field:
![image](https://github.com/transmitterdan/VDRplayer/assets/55266910/9c59b06b-b929-4bdc-976b-3d55e1237c90)

Or is the window you showed in reply not OpenCPN at all?

Anyway, it's working fine.  As I am recording ShipDriver Plugin tracks at 50-100 kt, playback  speed is not an issue.  

And I've now created a race triangular course route that doubles around itself twice, 6 legs around 3 waypoints.  Shipdriver follows it faithfully with no confusion over the repeats and your VDRplayer replays it perfectly.  This has been often buggy on actual racing, or maybe it has been me, Navigator.  But doing these simulations is very helpful and confidence building.  

Thanks again to you and Gleason.
Alan

**Comment by AlanH70** [comment]
Specing 127.0.0.0 did not work:
```
PS C:\Users\alandell3000\Desktop> Python C:\Users\alandell3000\AppData\Local\Programs\Python\Python312\Scripts\VDRplayer.py --port=2947 --sleep=.07 --UDP --dest=127.0.0.0 Hakefjord.txt
Playing file 'Hakefjord.txt', Type Ctrl-C to exit...
  UDP target IP: 127.0.0.0
UDP target port: 2947
Inserting 70.00 mS delay between each message.

Exception...
[WinError 10051] A socket operation was attempted to an unreachable network
```
But removing the command line UDP destination, allowing default to whatever localhost was did work, revealing the true IP of localhost 192.168.0.14:
```
PS C:\Users\alandell3000\Desktop> Python C:\Users\alandell3000\AppData\Local\Programs\Python\Python312\Scripts\VDRplayer.py --port=2947 --sleep=.07 --UDP --fast=10.0 Hakefjord.txt
Playing file 'Hakefjord.txt', Type Ctrl-C to exit...
  UDP target IP: 192.168.0.14
UDP target port: 2947
Inserting 70.00 mS delay between each message.
 60.5 percent complete....
KeyboardInterrupt.
Exiting cleanly.
```
As the track file was based on a 5 kt vessel, runs well but slow, and my attempt to speed up with the --fast=10 spec did not speed it up.
Is that because the NMEA file is not NMEA 4 compliant?  What is the difference?    
Also, noted specifying in OpenCPN the IP of 127.0.0.0 or the 192.168 both worked - did not seem to matter as the correct socket was set there.

**Reply by transmitterdan** [reply]
I guess your file is not a V4 file.  Try instead --speed=0.07 and see what happens.

**Reply by AlanH70** [reply]
The file is your "Hakefjord.txt" from the github files.  
I'll try the 0.07 spec next.

**Reply by AlanH70** [reply]
I dropped a MOB mark as a visual reference.  I do not see --speed= as a command line option, so I varied --fast- from 0.07 to1.0 to 200000.0 and the boat looks to not change its onscreen speed.

I'll try --speed= next

**Reply by AlanH70** [reply]
Nope 
`option --speed not recognized`

**Comment by transmitterdan** [comment]
I am testing OcpnCPN 5.10 so that's how it will appear when released.  But I think you see what is needed for the older version.

**Reply by AlanH70** [reply]
Ah, got it. V5.10  

Yes, I believe i do see.  Success!

**Comment by transmitterdan** [comment]
My apologies.  Try --sleep=0.07

**Reply by AlanH70** [reply]
Oh, I've been running --sleep=.07, but you add a leading zero.  I'll try "--sleep=0.07"

No joy.  Both sleep specs result in script response of 70 msec delay between messages and boat is just as slow on screen
```
PS C:\Users\alandell3000\Desktop> Python C:\Users\alandell3000\AppData\Local\Programs\Python\Python312\Scripts\VDRplayer.py --port=2947 --sleep=0.07 --UDP --fast=10.0 Hakefjord.txt
Playing file 'Hakefjord.txt', Type Ctrl-C to exit...
  UDP target IP: 192.168.0.14
UDP target port: 2947
Inserting 70.00 mS delay between each message.
```

**Reply by transmitterdan** [reply]
0.07 is 70mS.  You can go lower which will play faster.  Try 0.03 or 0.02.  Even 0.01 should work.

**Reply by AlanH70** [reply, marked as answer]
Yes, and at 1 msec, --sleep=0.001, it really flys!  It completes the entire Swedish route in 15 seconds or less, and, as it says, --fast is ignored  and I see no difference between fast=1 and fast=10 with sleep at 1 msec.

**Comment by transmitterdan** [comment]
Fast is useless for NMEA3 files such as this. To control playout speed set the delay between sentences using the --sleep option.
